Drop Ruby < 2.3 support

pirj opened the pull request:
Better viewed ignoring indentation changes https://github.com/rspec/rspec-support/pull/436/files?diff=split&w=1

As per https://github.com/rspec/rspec/issues/61#issuecomment-687290045


Sibling PRs:
 - https://github.com/rspec/rspec-mocks/pull/1349
 - https://github.com/rspec/rspec-expectations/pull/1231
 - https://github.com/rspec/rspec-core/pull/2787


Minimal changes to `rspec-core` and `rspec-mocks` were required to run `rspec-support`'s own spec suite, pushed to branches with the same name.

## What has been done

Done:
- [x] clean up RubyFeatures
- [x] check respond_to?
- [x] check method_defined?
- [x] check send (where we call private methods that later became public)
- [x] cleanup Gemfile
- [x] update required_ruby_version
- [x] update VERSION
- [x] check RUBY_VERSION usages
- [x] check if we require_rspec_support ruby_features where RubyFeatures are not used
- [-] check define_method where def would suffice
- [x] check `(class << self; self; end)` usages for possibility to use `singleton_class` (some can't be replaced as `singleton_class` does create a singleton class if it doesn't exist yet)
- [x] check if Java::JavaLang::String.instance_method(:char_at).arity == -1
- [?] check Java::JavaLang::String.instance_method(:char_at).parameters == [] - still the case
- [x] check 1.8, 1.9, 2.0, 2.1, jruby, 9.0, 9.1, mri
- [x] check requires for support's mutex
- [x] check `eval`
- [x] check "work around", "workaround" - there are two cases: check if this was fixed in JRuby 9.2
- [x] check docs
- [-] ~WithKeywordsWhenNeeded~ - will be addressed separately
- [x] `gsub`/`unindent` where squiggly heredoc can be used
- [-] ~remove `ripper_supported?`~ - it has poor support in JRuby 9.1

pirj committed:
Simplify fork_supported?

pirj committed:
Simplify ripper_supported?

pirj committed:
Remove redundant jruby_9000?

We always test against 9.0+, 9.2 actually, and 9.2 is recommended for
all users, there's no excuse to be using 9.0 or 9.1

pirj committed:
Remove kw_args_supported?

pirj committed:
Remove supports_rebinding_module_methods?

pirj committed:
Remove required_kw_args_supported?

pirj committed:
Always use require_relative

pirj committed:
Remove optional_and_splat_args_supported?

pirj committed:
Remove caller_locations_supported?

pirj committed:
Remove supports_exception_cause?

pirj committed:
Remove unused module_refinement_supported?

pirj committed:
Remove JRuby arity check workaround

Java::JavaLang::String.instance_method(:char_at).arity evaluates to 1 on
JRuby 9.2(.13.0)

pirj committed:
Remove 1.8.7-specific workaround

pirj committed:
Remove JRuby in 1.8.7 compat mode workaround

pirj committed:
Remove 1.9 Open3 workaround

pirj committed:
Remove module_prepends_supported?

pirj committed:
Remove String#encoding workarounds

pirj committed:
Remove Time#nsec workaround

pirj committed:
Remove 1.9.2-specific workaround

pirj committed:
Simplify a macro definition

pirj committed:
Update required Ruby version gemspec constraint

pirj committed:
Remove explicit gem version constraints

pirj committed:
Remove assorted older rubies workarounds

pirj committed:
Explicitly require ruby_features where needed

pirj committed:
Remove const_defined? workaround

pirj committed:
Remove version-specific details from a doc

pirj committed:
Remove old JRuby workaround

pirj committed:
Remove usages of jruby_9000

pirj committed:
Remove non_mri?

pirj commented:
For some reason, builds stopped triggering.
Going to switch to "ready for review" temporarily just to figure out if it's the case (quite unlikely though).

pirj committed:
Add Changelog entry

pirj committed:
Get rid of evals

benoittgt commented:
For the build issue it is because travisci config.  https://github.com/rspec/rspec-support/pull/436/commits/112973040cbab3bfb69b23bc21b03bc084c93e00

JonRowe commented on .rubocop_rspec_base.yml:
```diff
@@ -39,28 +39,12 @@ CyclomaticComplexity:
 
-# We still support 1.8.7 which requires trailing dots
-DotPosition:
-  EnforcedStyle: trailing
```
We should leave this rule to avoid churn

JonRowe reviewed (commented):
Theres a whole bunch of unnecessary changes, I don't feel we need to switch to 1.9 style keys just because we can, similarly with the trailing dot changes.

I also think we should keep the mutex method binding, the reasoning behind it is sound, we just don't need the vendored implementation.

pirj commented:
@JonRowe Thanks for the review.

I've outlined the things to discuss in the description, including:
> [?] extract rubocop changes into a separate pull request

I'll extract this into a separate follow-up pull request to `4-0-dev`.

pirj commented on lib/rspec/support/reentrant_mutex.rb:
```diff
@@ -40,22 +36,5 @@ def exit
-      class Mutex < ::Mutex
-        # If you mock Mutex.new you break our usage of Mutex, so
-        # instead we capture the original method to return Mutexs.
-        NEW_MUTEX_METHOD = Mutex.method(:new)
```
Fair enough. Reminds me of the recent case with stubbed `File.read` 👍

pirj commented on lib/rspec/support/spec/shell_out.rb:
```diff
@@ -62,13 +45,6 @@ def run_ruby_with_current_load_path(ruby_command, *flags)
-          %r{lib/ruby/stdlib/jar},
-          # This is a JRuby file that generates warnings on 9.1.7.0
-          %r{org/jruby/RubyKernel\.java},
-          # This is a JRuby gem that generates warnings on 9.1.7.0
```
I remember @headius mentioned that for the end-users there is no good excuse for not using the latest 9.2 instead of any other JRuby version. Did I understand/remember this correctly?

pirj commented on lib/rspec/support/method_signature_verifier.rb:
```diff
@@ -177,37 +144,6 @@ def classify_parameters
-      # This workaround only works for Java proxy methods, and in order to
-      # support regular methods and blocks, we need to be careful about calling
-      # owner and java_class as they might not be available
-      if Java::JavaLang::String.instance_method(:char_at).arity == -1
```
This is fixed in jruby-9.2.13.0, I'll double-check with earlier 9.2 versions.

JonRowe replied:
Did you double check 9.1.x.x.?

pirj replied:
I did, but with 9.1.17.0.

WDYT of temporarily adding an earlier version to CI, like e.g. 9.1.7.0 that we had previously?

pirj replied:
`Java::JavaLang::String.instance_method(:char_at).arity` resolves to `1` in all of 9.2.13.0, 9.1.17.0, 9.1.7.0, and even 9.1.0.0.

So it's not like that it's fixed, but more like we don't support versions anymore that had that issue.

pirj commented on lib/rspec/support/ruby_features.rb:
```diff
@@ -28,22 +27,10 @@ def jruby?
-      def non_mri?
-        !mri?
-      end
-
```
I'm not sure if you specifically mean `non_mri?`? Couldn't find `non_mri` in git history in Mocks/Expectations/Core/Support.

JonRowe replied:
Its intended to be available for us to write specs with when needed.

pirj replied:
The most recent usage of it was ~10 years ago.
When writing specs we can use `unless: mri?` in place of `if: non_mri?`, does it do the job?

pirj commented on lib/rspec/support/with_keywords_when_needed.rb:
```diff
@@ -8,26 +8,19 @@ module WithKeywordsWhenNeeded
-      else
-        def class_exec(klass, *args, &block)
+      # Remove this in RSpec 4 in favour of explictly passed in kwargs where
+      # this is used. Works around a warning in Ruby 2.7
```
Sure, it's in TODO in the description:

> [ ] WithKeywordsWhenNeeded

pirj commented on rspec-support.gemspec:
```diff
@@ -33,8 +33,8 @@ Gem::Specification.new do |spec|
+  spec.required_ruby_version = '>= 2.3.0'
 
-  spec.add_development_dependency "rake",         "> 10.0.0"
+  spec.add_development_dependency "rake", ">= 12.3.3"
```
I preferred to avoid updating libraries versions, [`12.3.3` comes from the `Gemfile`](https://github.com/rspec/rspec-support/blob/main/Gemfile#L21):
```ruby
gem 'rake', '>= 12.3.3'
```

pirj commented on script/functions.sh:
```diff
@@ -12,10 +12,8 @@ SPECS_HAVE_RUN_FILE=specs.out
-  export RUBYOPT="--disable=gem"
-fi
+# without the extra load time of rubygems.
+export RUBYOPT="--disable=gem"
```
Sure, that's part of the iterative plan.

> for rspec-dev:
>   remove unused predicate_functions

I'm experimenting here to figure out what can be stripped off, so we can find what needs to remain because it's still used in some of the repos and what can be removed in `rspec-dev`. I'll extract this and will send a follow-up PR to `rspec-dev` and to other repos against `4-0-dev`.

pirj commented on spec/rspec/support/mutex_spec.rb:
```diff
@@ -1,8 +0,0 @@
-    expect(Mutex).to receive(:new)
-    ::Mutex.new
-  end
-end
```
Ooops!

pirj commented:
@JonRowe @benoittgt Appreciate if you can take a look at Discuss section of the description, there are some important questions, let me cite below:

1. introduce https://github.com/ruby/ruby2_keywords dependency and remove respond_to?(:ruby2_keywords, true) check
2. cleanup of build functions/predicates - any objections? (for sure in a separately pull request, and coming from rspec-dev)
3. drop Ruby 2.3 - problematic'ish due to a dependency on outdated OpenSSL, EOL'ed, not supported by RuboCop, still used by 8% of users
4. tweak clone_repo to accept a branch, use maintenance-branch's branch for main repos, and 4-0-maintenance for cloning rspec-rails since it's alienated anyway?

JonRowe commented:
> [?] introduce https://github.com/ruby/ruby2_keywords dependency and remove respond_to?(:ruby2_keywords, true) check

No, we can't introduce dependencies.

> [?] drop Ruby 2.3 - problematic'ish due to a dependency on outdated OpenSSL, EOL'ed, not supported by RuboCop, still used by 8% of users

What Rubocop supports does not drive us. 2.3 is not really problematic.

> [?] tweak clone_repo to accept a branch, use maintenance-branch's branch for main repos, and 4-0-maintenance for cloning rspec-rails since it's alienated anyway?

It already does this? Its just hardcoded based on the branch?

> [?] extract rubocop changes into a separate pull request

Just remove them, they're unneeded.

pirj commented:
> >    and 4-0-maintenance for cloning rspec-rails since it's alienated anyway?

> It already does this? Its just hardcoded based on the branch?

Doesn't seem to. The build was failing without `drop-old-rubies` `rspec-rails` branch. Also, it failed with default `version.rb`:

```
Resolving dependencies...

Bundler could not find compatible versions for gem "rspec-core":

  In Gemfile:
    rspec-core
    rspec-rails was resolved to 4.1.0.pre, which depends on
      rspec-core (= 3.11.0.pre)
```

Changed it to `4.0.0.pre` to trick this build.

pirj committed:
Fix offences

pirj commented:
I'm getting stuck with removing `WithKeywordsWhenNeeded` and dealing with kwargs delegation related failures.
Removal of `WithKeywordsWhenNeeded` is tangential to the removal of rubies, and I'm going to remove all `eval` hacks, but not kwargs delegation hacks.
I intend to address the removal of `WithKeywordsWhenNeeded` and kwargs delegation issues and hacks in a separate series of pull requests.

pirj committed:
Use Mutex from core, prevent it from being stubbed

pirj commented on .github/workflows/ci.yml:
```diff
@@ -26,9 +26,11 @@ jobs:
           - 2.5
           - 2.4
           - 2.3
+          - jruby-9.2
```
This worked fine https://github.com/rspec/rspec-support/pull/436/checks?check_run_id=1418962579

Are you interested in incorporating this in your GHA initiative?

I've removed `JRUBY_OPTS='--dev'` with no replacement for GHA. Not sure if it's needed.

The reason I'm doing this in this PR (temporarily) is to test this out while Travis is 🐌 
```
      # To work around JRuby error:
      # can't convert RSpec::Support::StdErrSplitter into String
```

pirj replied:
Sorry, forgot to mention @JonRowe ☝️

pirj replied:
`jruby-head` failure seems to be due:

```
Warning: the --server flag is deprecated and has no effect most JVMs
```

[This will be fixed in 9.3.0.0](https://github.com/jruby/jruby/issues/6147)

Wondering why it doesn't affect our `jruby-9.2` 😕

pirj replied:
And https://github.com/rspec/rspec-support/commit/33e6318 didn't help, since it outputs to stdout 😞

JonRowe replied:
I'd leave this the same as main but remove the 2 old rubies.

pirj committed:
Remove StdErrSplitter workarounds

Kept gems warning muted due to:

1.

      RuntimeError:
       Warnings were generated: /home/runner/work/rspec-support/bundle/ruby/2.5.0/gems/actioncable-6.0.3.4/lib/action_cable/channel/test_case.rb:178: warning: method redefined; discarding old connection
     Shared Example Group: "an rspec-rails example group mixin" called from ./spec/rspec/rails/example/channel_example_group_spec.rb:6
     # /home/runner/work/rspec-support/rspec-support/lib/rspec/support/spec/stderr_splitter.rb:38:in `verify_no_warnings!'

2.
       RuntimeError:
        Warnings were generated: /home/runner/work/rspec-support/bundle/ruby/2.5.0/gems/actionpack-6.0.3.4/lib/action_dispatch/routing/route_set.rb:560: warning: instance variable @_routes not initialized
      Shared Example Group: "an rspec-rails example group mixin" called from ./spec/rspec/rails/example/view_example_group_spec.rb:5
      # /home/runner/work/rspec-support/rspec-support/lib/rspec/support/spec/stderr_splitter.rb:38:in `verify_no_warnings!'

pirj commented:
@JonRowe I intend to add back a `jruby-9.1` build job, and make an attempt to remove `ripper_supported?` after that.

JonRowe commented on lib/rspec/support/reentrant_mutex.rb:
```diff
@@ -40,22 +38,5 @@ def exit
-      class Mutex < ::Mutex
-        # If you mock Mutex.new you break our usage of Mutex, so
-        # instead we capture the original method to return Mutexs.
-        NEW_MUTEX_METHOD = Mutex.method(:new)
```
Just restore this and remove the other file, we don't need the file just this.

JonRowe replied:
I still feel it would be better to roll the other file in here.

JonRowe commented on lib/rspec/support/spec/stderr_splitter.rb:
```diff
@@ -28,30 +26,12 @@ def reopen(*args)
-
-        # Ruby 2.7.0 complains about hashes used in place of keyword arguments
-        # Aruba 0.14.2 uses this internally triggering that here
-        return if line =~ %r{lib/ruby/2\.7\.0/fileutils\.rb:622: warning:}
```
Restore these, we support all versions of 2.7.

JonRowe replied:
Still this.

JonRowe commented on lib/rspec/support/spec/stderr_splitter.rb:
```diff
@@ -28,30 +26,12 @@ def reopen(*args)
-        # Aruba 0.14.2 uses this internally triggering that here
-        return if line =~ %r{lib/ruby/2\.7\.0/fileutils\.rb:622: warning:}
+        # Ignore warnings coming from gems, specifically Rails on Ruby 2.5+
+        return if line =~ %r{^\S+/gems/\S+:\d+: warning:}
```
Don't change this logic when its unrelated, we don't want to become less specific.

JonRowe commented on .travis.yml:
```diff
@@ -15,27 +15,14 @@ before_install:
+    - rvm: 2.7.2
       env: DIFF_LCS_VERSION="~> 1.3.0"
-    - rvm: 2.7.1
+    - rvm: 2.7.2
```
Don't need to bother changing these, they will be on github actions soon enough.

JonRowe commented on .travis.yml:
```diff
@@ -15,27 +15,14 @@ before_install:
 bundler_args: "--binstubs --standalone --without documentation --path ../bundle"
 script: "script/run_build"
 rvm:
-  - 1.8.7
```
Really you can just delete this file, we won't be using travis for RSpec 4's CI.

pirj replied:
That's for sure, once it's all done.
But before it was done, I preferred to have at least some builds to constantly updating two repos I was actively committing to to `drop-old-rubies` to those frequent commits to `4-0-dev`.
Once it's all settled, I'll go through all the changes thoroughly again myself and will ask for a detailed review.

pirj committed:
Remove redundant boilerplate

pirj commented on lib/rspec/support/spec/shell_out.rb:
```diff
@@ -16,26 +16,9 @@ def with_env(vars)
 
-      if Open3.respond_to?(:capture3) # 1.9+
-        def shell_out(*command)
-          stdout, stderr, status = Open3.capture3(*command)
```
I'm not sure how this used to work.
jruby-9.1 (9.1.17.0):
```
> require 'open3'
 => true
jruby-9.1.17.0 :003 > Open3.respond_to?(:capture3)
 => true
jruby-9.1.17.0 :004 > o,e,s=Open3.capture3("ruby -e 'warn \"lol\"'")
lol
 => ["", "", #<Process::Status: pid 12445 exit 0>]
```

jruby-9.2 (9.2.13.0), MRI 2.7.1:
```
> require 'open3'
=> true
[2] pry(main)> o,e,s=Open3.capture3("ruby -e 'warn \"lol\"'")
=> ["", "lol\n", #<Process::Status: pid 12883 exit 0>]
```

Can't find anything related to `capture3` in JRuby's changelog.
I was going to add a check for JRuby < 9.2 and use `popen3`, but:
```
jruby-9.1.17.0 :006 > Open3.popen3("ruby -e 'puts \"lol\"'") do |_,o,e|
jruby-9.1.17.0 :007 >     puts "ERR: #{e.read}", "OUT: #{o.read}"
jruby-9.1.17.0 :008?>   end
lol
ERR:
OUT:
 => nil
```

also, there were two `popen3` related issues fixed in 9.1 and 9.2:
#5235 - popen3 doesn't works fine with java 9/10 and jruby > 1.9.13
#3948 - Open3.popen3 with block fails on Windows
We can't discard Java 9/10 and Windows, can we?

Locally, I'm getting this with 9.1.17.0:
```
  1) rspec warnings and deprecations works when required in isolation
     Failure/Error: DEFAULT_FAILURE_NOTIFIER = lambda { |failure, _opts| raise failure }
       expected "" to start with "DEPRECATION: foo is deprecated"

Failed examples:

rspec ./spec/rspec/support/warnings_spec.rb:11
```
This also fails in [GHA](https://github.com/rspec/rspec-support/runs/1438630956?check_suite_focus=true), and locally on `4-0-dev`, and `main`, with both my changes to other repos and other RSpec gems from RubyGems.

`USE_GIT_REPOS=1` doesn't work with Bunlder 2:
```
NoMethodError: undefined method `[]' for nil:NilClass
  /Users/pirj/.rvm/rubies/jruby-9.1.17.0/lib/ruby/gems/shared/gems/bundler-2.1.4/lib/bundler/source/git/git_proxy.rb:90:in `version'
  /Users/pirj/.rvm/rubies/jruby-9.1.17.0/lib/ruby/gems/shared/gems/bundler-2.1.4/lib/bundler/source/git/git_proxy.rb:147:in `block in copy_to'
  org/jruby/RubyDir.java:323:in `chdir'
```

Yet, still, with Bundler 1 and `rspec-*` from repos, the failure remains.

So, on `main`, with git repos, on JRuby 9.1.17.0, I'm getting:
```
520 examples, 3 failures, 5 pending

Failed examples:

rspec ./spec/rspec/support/spec/shell_out_spec.rb:6 # RSpec::Support::ShellOut shells out and returns stdout and stderr
rspec ./spec/rspec/support/spec/shell_out_spec.rb:33 # RSpec::Support::ShellOut passes along the provided ruby flags
rspec ./spec/rspec/support/warnings_spec.rb:11 # rspec warnings and deprecations works when required in isolation
```

which are exactly the same failures I'm getting with my changes on CI.

Is this something you've seen along your journey with GHA, @JonRowe, @benoittgt ?

pirj replied:
Just in case:
```
$ ruby --version
jruby 9.1.17.0 (2.3.3) 2018-04-20 d8b1ff9 OpenJDK 64-Bit Server VM 14.0.1+14 on 14.0.1+14 +jit [darwin-x86_64]
```

JonRowe replied:
Er. Pass.

pirj committed:
Remove BasicObject checks

pirj commented on lib/rspec/support/ruby_features.rb:
```diff
@@ -103,88 +60,20 @@ def supports_taint?
+      # or cannot parse source including `:if`.
+      # Ripper on JRuby 9.x.x.x < 9.1.17.0 can't handle keyword arguments
+      # Neither can JRuby prior to 9.2.1.0
+      if Ruby.rbx? || (Ruby.jruby? && JRUBY_VERSION < '9.2.1.0')
```
@JonRowe, do we keep Rubinius support, and this check here, and just dropping it from our CI matrix? Or should I remove this check, and only keep JRuby check?

JonRowe replied:
Keep the check, and just drop from our CI matrix. If someone comes to us with Rubinius or Truffle problems we will try to help, just not going to run CI agains them, put back the call to `jruby_version` though :)

pirj committed:
Remove filtering for unsupported rubies

pirj committed:
Fix code doc

pirj commented:
No CI-related files included in this change!
Suddenly an unexpectedly all-green 🎉 
@benoittgt @JonRowe welcome for a review.

pirj commented on maintenance-branch:
```diff
@@ -1 +1 @@
-4-0-dev
\ No newline at end of file
+drop-old-rubies
```
⚠️ This change is temporary and needs to be removed before merging, or soon after ⚠️ 

This is due to changes from sibling PRs being inter-dependent, e.g. a removed `jruby_9000?` method is still in use in `rspec-core` on branch 4-0-dev, but all its usages are removed on branch drop-old-rubies.

JonRowe replied:
We'll be merging this PR last, so its should be safe to drop this.

pirj replied:
I can't drop this one - other repos' 4-0-dev branches still use methods that we've removed here (`jruby_version` is one example 😉 ). If we run their corresponding build steps, their specs will fail.

Let's remove this change once we've merged the rest of the repos?

JonRowe replied:
Other repos are merged, do you want to drop this change?

JonRowe commented on Changelog.md:
```diff
@@ -1,3 +1,7 @@
+# Development (unreleased)
+
+* Drop support for Ruby < 2.3. (Phil Pirozhkov, #436)
```
```suggestion
## Breaking Changes

* Ruby < 2.3 is no longer supported. (Phil Pirozhkov, #1349)
```

JonRowe commented on Gemfile:
```diff
@@ -13,51 +13,21 @@ branch = File.read(File.expand_path("../maintenance-branch", __FILE__)).chomp
-  gem "childprocess", ">= 3.0.0"
-end
+gem "childprocess", ">= 3.0.0"
+gem 'ffi', '~> 1.13.0'
```
This is incorrect, it needs 1.12 until when theres no Rubygems usage?

pirj replied:
Again, I left this intact 😆 

![image](https://user-images.githubusercontent.com/6916/101819428-4bbecb80-3b36-11eb-8aa9-9b35980cafeb.png)

Will fix this across repos 👍

JonRowe commented on lib/rspec/support.rb:
```diff
@@ -12,22 +12,13 @@ module Support
+      define_singleton_method("require_rspec_#{lib}") do |f|
+        require_relative.call("#{lib}/#{f}")
       end
     end
```
:heart:

JonRowe commented on lib/rspec/support/comparable_version.rb:
```diff
@@ -1,46 +0,0 @@
-module RSpec
```
Restore this please, it's used for better comparison of the 4 digit JRuby strings.

JonRowe commented on lib/rspec/support/mutex.rb:
```diff
@@ -1,73 +1,11 @@
+    class Mutex < ::Mutex
+      class << self
+        define_method(:new, &::Mutex.method(:new))
       end
```
💜

JonRowe commented on lib/rspec/support/ruby_features.rb:
```diff
@@ -28,22 +27,10 @@ def jruby?
 
-      def jruby_version
-        @jruby_version ||= ComparableVersion.new(JRUBY_VERSION)
-      end
```
Restore this.

JonRowe commented on spec/rspec/support/comparable_version_spec.rb:
```diff
@@ -1,39 +0,0 @@
-require 'rspec/support/comparable_version'
```
Restore.

JonRowe commented on spec/rspec/support/ruby_features_spec.rb:
```diff
@@ -151,21 +112,17 @@ def ripper_can_parse_source_referencing_keyword_arguments?
         end
 
         it 'returns whether Ripper is correctly implemented in the current environment' do
-          if RSpec::Support::Ruby.jruby? && RSpec::Support::Ruby.jruby_version < '9.2.1.0'
```
Restore jruby_version check.

JonRowe reviewed (changes requested):
Looking good overall, some comments still to be addressed and some stuff to put back.

pirj committed:
Downgrade ffi due to their usage of Gem

pirj commented:
Updates:
 - reverted `jruby_version` removal and its usages
 - pinned `ffi` at `~> 1.12.0`
 - accepted Changelog proposed change (breaking change section)
 - reverted error filtering changes in stderr_splitter

pirj commented:
Update: combined ReentrantMutex with Mutex.

JonRowe commented on lib/rspec/support/recursive_const_methods.rb:
```diff
@@ -5,52 +5,27 @@ module Support
-        def const_defined_on?(mod, const_name)
-          mod.const_defined?(const_name, false)
-        end
+      #   C.const_get("Hash") # => ::Hash
```
```suggestion
      # Then:
      #   C.const_get("Hash") # => ::Hash
```

JonRowe commented on lib/rspec/support/recursive_const_methods.rb:
```diff
@@ -5,52 +5,27 @@ module Support
-      # we need to conditionally define methods to ignore the top-level/inherited
-      # constants.
       #
-      # Given:
```
I think it reads better if this goes back.

pirj replied:
Agree, it is non-trivial, and additional verbosity won't hurt.

JonRowe reviewed (changes requested):
Almost there, just the branch pinning to drop and see if it builds :joy:

pirj commented:
Update: reset the temporary `maintenance-branch` commit.

Green! 💚

pirj commented:
🎉
